src/ValueParsers/TimeParser.php

```diff
@@ -43,83 +44,113 @@ class TimeParser extends StringValueParser {
+	 * @return string[] Array with index 0 => sign, 1 => year, 2 => month, 3 => day, 4 => hour,
+	 * 5 => minute, 6 => second and 7 => calendar model.
 	 */
-	private function padYear( $year ) {
-		return str_pad( $year, 16, '0', STR_PAD_LEFT );
+	private function splitTimeString( $value ) {
+		$pattern = '@^\s*'                                                //leading spaces
+			. "([-+\xE2\x88\x92]?)\\s*"                                   //sign
+			. '(\d{1,16})-(\d{2})-(\d{2})'                                //year, month and day
+			. '(?:T(\d{2}):(\d{2})(?::(\d{2}))?)?'                        //hour, minute and second
+			. 'Z?'                                                        //time zone
+			. '\s*\(?\s*' . CalendarModelParser::MODEL_PATTERN . '\s*\)?' //calendar model
+			. '\s*$@iu';                                                  //trailing spaces
+
+		if ( !preg_match( $pattern, $value, $matches ) ) {
+			throw new ParseException( 'Malformed time', $value, self::FORMAT_NAME );
+		} elseif ( $matches[2] < 60 && $matches[5] === '' ) {
+			throw new ParseException( 'Not enough information to decide if the format is YY-MM-DD',
+				$value, self::FORMAT_NAME );
+		} elseif ( $matches[3] > 12 ) {
+			throw new ParseException( 'Month out of range', $value, self::FORMAT_NAME );
+		} elseif ( $matches[4] > 31 ) {
+			throw new ParseException( 'Day out of range', $value, self::FORMAT_NAME );
+		} elseif ( $matches[5] > 23 ) {
+			throw new ParseException( 'Hour out of range', $value, self::FORMAT_NAME );
+		} elseif ( $matches[6] > 59 ) {
+			throw new ParseException( 'Minute out of range', $value, self::FORMAT_NAME );
+		} elseif ( $matches[7] > 62 ) {
+			throw new ParseException( 'Second out of range', $value, self::FORMAT_NAME );
+		}
+
+
+		$matches = array_slice( $matches, 1 );
+		$matches[0] = str_replace( "\xE2\x88\x92", '-', $matches[0] );
+
+		return $matches;
 	}
 
 	/**
-	 * @param array $timeParts with the following keys.
-	 *            sign, year, month, day, hour, minute, second, calendar
+	 * @param string[] $timeParts Array with index 0 => sign, 1 => year, 2 => month, etc.
 	 *
-	 * @return int precision as a TimeValue PRECISION_ constant
+	 * @return int Precision as a TimeValue::PRECISION_ constant
 	 */
-	private function getPrecisionFromTimeParts( $timeParts ) {
-		if( $timeParts['second'] !== '00' ) {
-			return TimeValue::PRECISION_SECOND;
-		}
-		if( $timeParts['minute'] !== '00' ) {
-			return TimeValue::PRECISION_MINUTE;
-		}
-		if( $timeParts['hour'] !== '00' ) {
-			return TimeValue::PRECISION_HOUR;
-		}
-		if( $timeParts['day'] !== '00' ) {
-			return TimeValue::PRECISION_DAY;
+	private function getPrecision( array $timeParts ) {
+		if ( intval( $timeParts[6] ) > 0 ) {
+			$precision = TimeValue::PRECISION_SECOND;
+		} elseif ( intval( $timeParts[5] ) > 0 ) {
+			$precision = TimeValue::PRECISION_MINUTE;
+		} elseif ( intval( $timeParts[4] ) > 0 ) {
+			$precision = TimeValue::PRECISION_HOUR;
```

Comment:
So, 215-02-15T00:00:00 would have precision Day, while 215-02-15T00:00:01 would have precision Second? Not sure I like that...  Wouldn't it be better if we could simply omit the "irrelevant" parts, instead of using 0 to indicate them? Or use * instead of 0.

It's my understanding that the input string for this parser isn't raw user input, but rather comes from other parsers or pre-parsers. So we'd be free to pick our own way to represent "don't care".

2015-02 <- precision month.
2015-02-** <- precision month.
2015-02-00T07 <- error
2015-02-**T07 <- error
2015-02-**T** <- precision month
2015-02-15T00:00 <- precision minute
2015-02-15T00:00:** <- precision minute
2015-02-15T00:00:00 <- precision second